jupyterhub/handlers/base.py

```diff
@@ -1470,20 +1474,31 @@ class UserRedirectHandler(BaseHandler):
 
     If the user is not logged in, send to login URL, redirecting back here.
 
+    If c.JupyterHub.user_redirect_hook is set, the return value of that
+    callable is used to generate the redirect URL.
+
     .. versionadded:: 0.7
     """
 
     @web.authenticated
-    def get(self, path):
-        user = self.current_user
-        user_url = url_path_join(user.url, path)
-        if self.request.query:
-            user_url = url_concat(user_url, parse_qsl(self.request.query))
+    async def get(self, path):
+        # If hook is present to generate URL to redirect to, use that instead
+        # of the default. The configurer is responsible for making sure this
+        # URL is right. If None is returned by the hook, we do our normal
+        # processing
+        url = None
+        if self.app.user_redirect_hook:
+            url = await self.app.user_redirect_hook(self, path)
```

Comment:
Do you think it's worth having `user` as an explicit argument to the hook? Rather than relying on the hook knowing the ins and outs of our BaseHandler API (which is a bit of a mess that I'd like to minimize exposing as public)? Further, what about being even more conservative and *only* passing `user, path` without exposing the Handler at all? What would be missing there that's useful? (first possible example: access to Authenticator instance)

Reply:
Good idea! I'm explicitly passing in user and the request object (Rather than just path), since you might wanna get access to the query string, etc. It does mean you need to deal with base_url and what not yourself - maybe we should pass 'app' as well?